src/hooks/useReportIDs.tsx

```diff
@@ -0,0 +1,135 @@
+};
+
+const ReportIDsContext = createContext<ReportIDsContextValue>({
+    orderedReportIDs: [],
+});
+
+function WithReportIDsContextProvider({
+    children,
+    chatReports,
+    betas,
+    policies,
+    allReportActions,
+    transactionViolations,
+    policyMembers,
+    priorityMode,
+    /**
+     * Only required to make unit tests work, since we
+     * explicitly pass the currentReportID in LHNTestUtils
+     * to SidebarLinksData, so this context doesn't have an
+     * access to currentReportID in that case.
+     *
+     * This is a workaround to have currentReportID available in testing environment.
+     */
+    currentReportIDForTests,
+}: WithReportIDsContextProviderProps) {
+    const currentReportIDValue = useCurrentReportID();
+    const derivedCurrentReportID = currentReportIDForTests ?? currentReportIDValue?.currentReportID;
+    const {activeWorkspaceID} = useActiveWorkspace();
+
+    const policyMemberAccountIDs = useMemo(() => getPolicyMembersByIdWithoutCurrentUser(policyMembers, activeWorkspaceID, getCurrentUserAccountID()), [activeWorkspaceID, policyMembers]);
+
+    const getOrderedReportIDs = useCallback(
+        (currentReportID?: string) =>
+            SidebarUtils.getOrderedReportIDs(
+                currentReportID ?? null,
+                chatReports,
+                betas ?? [],
+                policies,
+                priorityMode,
+                allReportActions,
+                transactionViolations,
+                activeWorkspaceID,
+                policyMemberAccountIDs,
+            ),
+        [chatReports, betas, policies, priorityMode, allReportActions, transactionViolations, activeWorkspaceID, policyMemberAccountIDs],
+    );
+
+    const orderedReportIDs = useMemo(() => getOrderedReportIDs(), [getOrderedReportIDs]);
+
+    // We need to make sure the current report is in the list of reports, but we do not want
+    // to have to re-generate the list every time the currentReportID changes. To do that
+    // we first generate the list as if there was no current report, then we check if
+    // the current report is missing from the list, which should very rarely happen. In this
+    // case we re-generate the list a 2nd time with the current report included.
+    const orderedReportIDsWithCurrentReport = useMemo(() => {
+        if (derivedCurrentReportID && !orderedReportIDs.includes(derivedCurrentReportID)) {
+            return getOrderedReportIDs(derivedCurrentReportID);
+        }
+        return orderedReportIDs;
+    }, [getOrderedReportIDs, derivedCurrentReportID, orderedReportIDs]);
+
+    const contextValue = useMemo(
+        () => ({
+            orderedReportIDs: orderedReportIDsWithCurrentReport,
+        }),
+        [orderedReportIDsWithCurrentReport],
+    );
+
+    return <ReportIDsContext.Provider value={contextValue}>{children}</ReportIDsContext.Provider>;
+}
+
+const ReportIDsContextProvider = withOnyx<WithReportIDsContextProviderProps, OnyxProps>({
```

Comment:
@hayata-suenaga  I agree with you and we shouldn't be wrapping with `withOnyx`. The reason I opted this approach was we need to access `reports`, `betas`, `policies` etc, in this context to carry out the calculations for `orderedReportIDs`. Also, I didn't find any usage like this in the codebase.

One solution to achieve the similar is to add `Onyx.connect` calls in this context for each of the Onyx key. 

Let me know whether we should refactor with `Onyx.connect`calls.

Reply:
> One solution to achieve the similar is to add Onyx.connect calls in this context for each of the Onyx key.

I suggest we pass the necessary data through the `ReportIDsContextProvider` props and delegate the Onyx subscription to the component where `ReportIDsContextProvider` is mounted.

I started a discussion on this topic in [[the open-source Slack channel](https://expensify.slack.com/archives/C01GTK53T8Q/p1710784425105419)](https://expensify.slack.com/archives/C01GTK53T8Q/p1710784425105419).

Reply:
>delegate the Onyx subscription to the component where ReportIDsContextProvider is mounted

Thanks for the starting the discussion. However, I guess this approach won't work because we won't be able to get the updated data to our context unless we are setting the state in the context as a callback from `SidebarlInksData`.

Also, if we decide to depend on `SidebarLinksData` for getting the Onyx data, we might not gain performance benefits. If you recall, by removing the dependency from `SidebarLinksData` in terms of generating the `reportIDs`, we benefited from the `reportIDs` being available to us each time the `SidebarlinksData` mounts.

Now, if we first wait for `SidebarLinksData` to give our context the updated data, this benefit will be lost.

Reply:
One thing we can also think of as a solution, is to consume the Onyx context from `OnyxProvider`. Then we can just call `useBetas(), useReportActions()` etc in our context. This way we will get rid of `withOnyx` HOC and still get the onyx data from the `OnyxProvider`.

Reply:
https://github.com/Expensify/App/blob/34fa9871c541ccfa03db46bc6e9e428288f718a3/src/components/OnyxProvider.tsx#L8-L45

Reply:
ah I was thinking we rather get the Onyx data in the `App.tsx` where the Provider is mounted.

However, I also appreciate that it might be weird if the root component subscribes to the Onyx data.

> One thing we can also think of as a solution, is to consume the Onyx context from OnyxProvider.

I also agree with you that having a context provider for Onyx is the natural solution, but this introduces a really big change. So, I think we won't go with this solution. (Even though there is a Onyx context provider, I probably won't call the hooks inside another provider to get the Onyx data).

Anyway, let's keep the current code.